vnext/Microsoft.ReactNative.Cxx/TurboModuleProvider.h

```diff
@@ -0,0 +1,94 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#pragma once
+#include <JSI/JsiAbiApi.h>
+#include <JSI/JsiApiContext.h>
+#include <ReactCommon/TurboModule.h>
+#include <winrt/Microsoft.ReactNative.h>
+#include <winrt/Windows.Foundation.h>
+
+namespace winrt::Microsoft::ReactNative {
+
+// CallInvoker on top of JSDispatcher.
+struct AbiCallInvoker : facebook::react::CallInvoker {
+  AbiCallInvoker(IReactDispatcher const &jsDispatcher) : m_jsDispatcher(jsDispatcher) {}
+
+  void invokeAsync(std::function<void()> &&func) override {
+    m_jsDispatcher.Post([func = std::move(func)]() { func(); });
+  }
+
+  void invokeSync(std::function<void()> &&func) override {
+    if (m_jsDispatcher.HasThreadAccess()) {
+      func();
+    } else {
+      std::mutex mutex;
+      std::condition_variable cv;
+      bool completed{false};
+      auto lock = std::unique_lock{mutex};
+      m_jsDispatcher.Post([&func, &mutex, &cv, &completed]() {
+        func();
```

Comment:
what if `func` throws?

Reply:
Good point! I have added catching and rethrowing it.